File: julia/4_draw_pascal.jl
# draw_pascal(n): a function that takes an integer n as a parameter and prints
# the first n rows of the Pascal's triangle.

function draw_pascal(n)
    # Check if n is a positive integer
    if n < 1 || n != floor(n)
        error("n must be a positive integer.")
    end

    # Initialize the first row of Pascal's triangle
    pascal_triangle = [[1]]

    # Generate Pascal's triangle
    for i in 2:n  # Start from the second row
        new_row = [1]  # Start with 1
        for j in 2:i-1  # 2nd to i-1 (inclusive). Fill the middle values
            new_row = push!(new_row, pascal_triangle[i-1][j-1]
                            + pascal_triangle[i-1][j])
        end
        push!(new_row, 1)  # End with 1
        push!(pascal_triangle, new_row)
    end

    # Print the triangle
    for row in pascal_triangle
        println(join(row, " "))
    end
end

# Example usage:
n = 5
draw_pascal(n)
# Output:
# 1
# 1 1
# 1 2 1
# 1 3 3 1
# 1 4 6 4 1

File: python/4_draw_pascal.py
# draw_pascal(N): a function that takes an integer N as a parameter and prints
# the first N rows of the Pascal's triangle.

# Leetcode Pascal's Triangle: https://leetcode.com/problems/pascals-triangle/

def draw_pascal(numRows):
    if numRows == 1:
        return [[1]]
    if numRows == 2:
        return [[1], [1, 1]]

    # if numRows > 2
    pascal_triangle = [[1], [1, 1]]

    for current_row in range(1, numRows - 1):  # range(1, 4)
        next_row_array = []
        for index in range(current_row + 1):  # range(2)
            if index == 0:
                next_row_array.append(1)
            else:
                next_row_array.append(pascal_triangle[current_row][index-1] + pascal_triangle[current_row][index])
        next_row_array.append(1)  # Add 1 to the end
        pascal_triangle.append(next_row_array)

    return pascal_triangle


# Example usage:
print(draw_pascal(5))
# [[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]

#     1
#    1 1
#   1 2 1
#  1 3 3 1
# 1 4 6 4 1